src/typesystem.cpp

```diff
@@ -100,11 +110,20 @@ TypeSystemCore::TypeSystemCore() {
 MonoId TypeSystemCore::new_term(
     TypeFunctionId tf, std::vector<int> args, char const* tag) {
 	tf = m_tf_core.find(tf);
-	int tf_data_idx = m_tf_core.find_function(tf);
-	int argument_count = m_type_functions[tf_data_idx].argument_count;
 
-	if (argument_count != -1 && argument_count != args.size())
-		assert(0 && "instanciating polymorphic type with wrong argument count");
+	{
+		// TODO: add a TypeFunctionTag::Unkown tag, to express
+		// that it's a dummy of unknown characteristics
+
+		// TODO: add some APIs to make this less jarring
+		TypeFunctionId dummy_tf =
+		    new_type_function(TypeFunctionTag::Builtin, {}, {}, true);
+
+		int dummy_tf_data_id = m_tf_core.find_function(dummy_tf);
+		m_type_functions[dummy_tf_data_id].argument_count = args.size();
+
+		m_tf_core.unify(tf, dummy_tf);
```

Comment:
So what this is saying is: we may not know what our type function actually is yet, but we do know the arguments we are instanciating this with right now. So we create a new dummy type function with that information so as to unify with other stuff and check for errors that way. Makes sense.

Reply:
exactly